### PR TITLE
Issue 41 add automated tests for cleaning file for all years

### PR DESCRIPTION
**Pull request by alexkcode**
# Description

This adds unit tests for Python script that cleans source data that contains building data for all years. The actual transformations tested should be explained by the title of each test.

Fixes # 41

# Testing Instructions

Run `python3 -m pytest test/data/scripts/unit/test_clean_all_years.py` from base project directory. All tests should pass as indicated by console pytest output.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: test/data/source/ChicagoEnergyBenchmarking.csv
Comment (vkoves) [inline test/data/source/ChicagoEnergyBenchmarking.csv:1]: Let's narrow this down - you can export in the data portal to query by certain things. My suggestion would be pick 20 or so buildings, 10 in particular that we care about (eg IIT, United Center, Art Institute) then a random 10 buildings and export from the city CSV

File: README.md [hunk 1]
```diff
@@ -100,3 +100,11 @@ cd src/data
 pip install --no-cache-dir -r requirements.txt
 bash run_all.sh
 ```
+### Run Data Processing Tests
+
+Run the following command for individual unit test suite (where XXX is something like 
+`test_clean_all_years`):
+
+```
+python3 -m pytest test/data/scripts/unit/XXX.py
```
Comment (vkoves) [inline README.md:109]: Is there a way to run all the tests?

Comment (alexkcode) [reply README.md:109]: I think I need to reconfigure somethings, pytest is not working from parent directory currently. Will update when working.

File: test/data/source/ChicagoEnergyBenchmarking.csv
Comment (alexkcode) [reply test/data/source/ChicagoEnergyBenchmarking.csv:1]: Will work on this

File: README.md (hunk 1 above)
Comment (alexkcode) [reply README.md:112]: running `pytest` at main project directory should now run all pytest tests

**Comment by alexkcode** [comment]
Current test cases only include buildings from `'United Center', 'Crown Hall', 'Art Institute'`

**Comment by alexkcode** [comment]
Current runtime is at around 21 seconds

**Comment by vkoves** [comment]
**Issue:** It looks like the actual data process now doesn't run, maybe due to the new imports?

**Comment by alexkcode** [comment]
@vkoves fixed all the pathing issues!

File: run_all.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# /bin/bash
+
+python3 -m src.data.scripts.clean_and_pare_down_data_all_years
+
+python3 -m src.data.scripts.process_data
+
+python3 -m src.data.scripts.add_context_by_property_type
+
+python3 -m src.data.scripts.clean_and_pare_down_data_current_year
```
Comment (vkoves) [inline run_all.sh:9]: Make sure there's a newline at the end of files.

File: src/data/scripts/clean_and_pare_down_data_all_years.py [hunk 3]
```diff
@@ -104,7 +105,11 @@ def main():
...
 
-    output_to_csv(all_recent_submitted_data, data_directory+data_out_file)
+    return all_recent_submitted_data
+
+def main():
+    processed = process(get_data_file_path(file_dir, building_emissions_file))
+    output_to_csv(processed, get_data_file_path(file_dir, data_out_file))
 
 if __name__ == "__main__":
     main()
```
Comment (vkoves) [inline src/data/scripts/clean_and_pare_down_data_all_years.py:115]: Add newline to EOF

File: src/data/scripts/clean_and_pare_down_data_current_year.py [hunk 4]
```diff
@@ -74,4 +74,5 @@
 
     # Mark columns as ints that should never show a decimal, e.g. Number of Buildings, Zipcode
     all_recent_submitted_data[int_cols] = building_data[int_cols].astype('Int64')
-    all_recent_submitted_data.to_csv(data_directory+data_out_file, sep=',', encoding='utf-8', index=False)
\ No newline at end of file
+    all_recent_submitted_data.to_csv(get_data_file_path(data_directory, data_out_file), 
+                                     sep=',', encoding='utf-8', index=False)
```
Comment (vkoves) [inline src/data/scripts/clean_and_pare_down_data_current_year.py:78]: Add newline to EOF

File: tests/data/scripts/unit/test_clean_all_years.py [hunk 5]
```diff
@@ -78,6 +92,14 @@ def test_int_values_remain_the_same(test_has_last_year_of_data):
...
     clean.output_to_csv(test_has_last_year_of_data, out_file)
     assert os.path.exists(out_file)
+
+@pytest.fixture
+def process():
+    return clean.process(get_file_path(src_dir, src_input_file))
+
+def test_data_has_ranking_columns(process):
+    for col in proc.building_cols_to_rank:
+        assert col in process.columns
```
Comment (vkoves) [inline tests/data/scripts/unit/test_clean_all_years.py:105]: Add newline to EOF

File: tests/data/source/test_src_data.csv [hunk 6]
```diff
@@ -0,0 +1,10 @@
+Data Year,ID,Property Name,Reporting Status,Address,ZIP Code,Chicago Energy Rating,Exempt From Chicago Energy Rating,Community Area,Primary Property Type,Gross Floor Area - Buildings (sq ft),Year Built,# of Buildings,Water Use (kGal),ENERGY STAR Score,Electricity Use (kBtu),Natural Gas Use (kBtu),District Steam Use (kBtu),District Chilled Water Use (kBtu),All Other Fuel Use (kBtu),Site EUI (kBtu/sq ft),Source EUI (kBtu/sq ft),Weather Normalized Site EUI (kBtu/sq ft),Weather Normalized Source EUI (kBtu/sq ft),Total GHG Emissions (Metric Tons CO2e),GHG Intensity (kg CO2e/sq ft),Latitude,Longitude,Location,Row_ID,Wards,Community Areas,Zip Codes,Census Tracts,Historical Wards 2003-2015
```
Comment (vkoves) [inline tests/data/source/test_src_data.csv:1]: How did you generate this file, and how can someone update it if the format of the CSV changes? Did you apply a certain set of queries to the city data?

File: tests/data/source/test_output.csv [hunk 7]
```diff
@@ -0,0 +1,3 @@
+DataYear,ID,PropertyName,ReportingStatus,Address,ZIPCode,ChicagoEnergyRating,ExemptFromChicagoEnergyRating,CommunityArea,PrimaryPropertyType,GrossFloorArea,YearBuilt,NumberOfBuildings,WaterUse,ENERGYSTARScore,ElectricityUse,NaturalGasUse,DistrictSteamUse,DistrictChilledWaterUse,AllOtherFuelUse,SiteEUI,SourceEUI,WeatherNormalizedSiteEUI,WeatherNormalizedSourceEUI,TotalGHGEmissions,GHGIntensity,Latitude,Longitude,Location,Row_ID,Wards,CommunityAreas,ZipCodes,CensusTracts,HistoricalWards2003-2015
```
Comment (vkoves) [inline tests/data/source/test_output.csv:1]: Same question here, if the format changed how would we update this?

**Comment by vkoves** [comment]
QA list for myself:

- [ ] Run tests locally
- [ ] Run data process, confirm output CSV doesn't change

**Comment by alexkcode** [comment]
Currently producing a slightly different output csv's than original script. Not sure why since the types and numbers are the same, but I'm still investigating.

**Comment by vkoves** [comment]
@alexkcode - looks like there's an issue creating the debug JSON, and some numeric columns become repeating numbers, it looks like it's the latitude and longitude column:

Screenshot from 2024-02-20 20-48-01

It's easiest to debug this by looking at the line by line diff, and then your editor will highlight the specific part of each line that changed.

**Comment by alexkcode** [comment]
After I changed the string type conversion in the script some of the columns like `ChicagoEnergyRating`  are now empty spaces, instead of Pandas type `nan`, in the produced `.csv` files. I think this is actually the correct behavior, but please let me know if it is not.

**Comment by alexkcode** [comment]
For specific string/column formats, we should check against the original source raw CSV (or as close as possible to it). I will change the `Latitude, Longitude` column rests and the other tests that test against the original values for the different direct type conversions that take place in the scripts.

**Comment by alexkcode** [comment]
Rendered websites (production vs. local) appear the same, i.e. the repeating decimals don't show up, unless my eyes missed something.

**Comment by alexkcode** [comment]
Could potentially be Pandas issue (from an update or something), will look into this.

**Comment by alexkcode** [comment]
Found decimal issue origin: before the conversion to string type in the Python scripts an interpretation of the data types is already by the `get_and_clean_csv` function.

**Comment by alexkcode** [comment]
@vkoves Please re-review!

**Comment by vkoves** [comment]
@alexkcode - I re-ran the data processing to make sure everything is up to date, but the test now fails with this one error:

Screenshot from 2024-03-12 20-16-04

I think that is related to the issue you filed, so I'll still merge this, since this isn't running in CI, but I'd love your help to fix that. I wonder if we could fix this just by rounding these floats to a certain relevant precision.

**Review by vkoves — COMMENTED** [review]
Looks good overall, let's ship this and improve it more later!

**Review by vkoves — APPROVED** [review]
Meant to approve but I had one open comment, so just fixed that. Let's ship!

**Pull request merged**